osquery/tables/applications/linux/docker_deb_packages.cpp

```
@@ -0,0 +1,198 @@
+  }
+
+  dpkg_teardown(&packages);
+  writer.EndArray();
+
+  size_t bufSize = buffer.GetSize();
+  if (bufSize < 1) {
+    return;
+  }
+  if (write(fd, buffer.GetString(), bufSize) != (ssize_t)bufSize) {
+    VLOG(1) << "error writing pkginfo ";
+  }
+}
+
+QueryData genDockerDebPackages(QueryContext& context) {
+  QueryData results;
+
+  for (const auto& id : context.constraints["id"].getAll(EQUALS)) {
+    if (!checkConstraintValue(id)) {
+      continue;
+    }
+
+    pt::ptree container;
+    Status s = dockerApi("/containers/" + id + "/json", container);
+    if (!s.ok()) {
+      VLOG(1) << "Error getting docker container info " << id << ": "
+              << s.what();
+      continue;
+    }
+    if (container.get<bool>("State.Running", false) != true) {
+      continue;
+    }
+    pid_t pid = container.get<int>("State.Pid", 0);
+    if (pid == 0) {
+      VLOG(1) << "Error getting pid for container " << id;
+      continue;
+    }
+
+    int fds[2];
+    if (socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == -1) {
+      VLOG(1) << "unable to open socket pair" << strerror(errno);
+      continue;
+    }
+
+    NamespaceOps nsOps(pid, fds[1]);
+    Status forkOps = nsOps.invoke(getDebPackages);
+
+    if (!forkOps.ok()) {
+      VLOG(1) << "error entering namespace: " << forkOps.what() << " : "
+              << forkOps.getCode();
+      close(fds[0]);
+      close(fds[1]);
+      continue;
+    }
+
+    close(fds[1]);
+    FILE* fp = fdopen(fds[0], "rb");
+    if (fp == NULL) {
+      VLOG(1) << "unable to open fd " << strerror(errno);
+      close(fds[0]);
+      Status k = nsOps.kill();
+      if (!k.ok()) {
+        VLOG(1) << "unable to terminate foked process " << k.what();
+      }
+      continue;
+    }
+    char readBuffer[65536];
+    rapidjson::FileReadStream is(fp, readBuffer, sizeof(readBuffer));
+    rapidjson::Document d;
+    d.ParseStream(is);
+    fclose(fp);
+    close(fds[0]);
+
+    Status w = nsOps.wait();
```

Note that wait() can fail early with EINTR due if a signal is delivered to the process; you probably want to restart the wait in that case.

---

Done

---

nit: Can this also be `auto wait = `?